raft.go

```diff
@@ -283,6 +283,22 @@ type Config struct {
 	// This behavior will become unconditional in the future. See:
 	// https://github.com/etcd-io/raft/issues/83
 	StepDownOnRemoval bool
+
+	// ResumeReplicateBelowPendingSnapshot allows the leader to resume replication
+	// to a follower (moving it to StateReplicate) if the follower is currently in
+	// StateSnapshot and we receive a MsgAppResp from it below the PendingSnapshot
+	// index but within the leader's log. In other words, the follower may apply a
+	// snapshot below the leader's PendingSnapshot, and the leader will resume
+	// replication as long as it connects the follower to the leader's log.
+	//
+	// Consider that complex systems may delegate the sending of snapshots to
+	// alternative datasources (i.e. not the leader). In such setups, it is
+	// difficult to manufacture a snapshot at a particular index requested by raft
+	// and the actual index may be ahead or behind. This should be okay, as long
+	// as the snapshot allows replication to resume.
+	//
+	// TODO(erikgrinaker): Consider making this the default behavior.
+	ResumeReplicateBelowPendingSnapshot bool
```

Comment:
If we are sure that (1) it will NOT cause any compatibility issue and (2) we will enable it by default and we don't expect/recommend users to disable it. Can we remove the option? We are adding more and more user facing options, and complicate the user interface.